brave/brave-ios
# Fix #6225: Ethereum Name Service (ENS) Navigation

## Summary of Changes
- Depends on IPFS support from #7013; this PR cannot merge without IPFS landing first.
- Refactored the SNS navigation logic out into `BraveWallet` target with a new `DecentralizedDNSHelper` class (w/ unit tests).
    - Enables us to re-use some logic, unit test, and lets us add new decentralized DNS services easier later (Unstoppable Domains!). 
    - Later we'll move the interstitital pages out into `BraveWallet` target too. This requires some refactoring of current logic (need to extract `TabContentScript` / `TabContentScriptLoader` / `InternalSchemeResponse` out of Client and into a shared target).
 - Added ENS support. ENS resolve method defaults to 'Ask' similar to existing ENS Offchain resolve method / SNS resolve method. When set to 'Ask' an interstitial page is shown before proceeding with the lookup.
 - Security/privacy review: https://github.com/brave/security/issues/1231

This pull request fixes #6225

## Submitter Checklist:


## Test Plan:
1. Open Web3 / Wallet settings and set 
    - `Resolve Ethereum Name Service (ENS) Domain Names` to `Ask`
    - `Allow ENS Offchain Lookup` to `Ask`
    - `Resolve IPFS Resources` to ask (optional); ENS will always load to an IPFS url; must be either `Ask` or `Enabled`
2. Navigate to any ENS domain via the omnibox. Ex. `offchainexample.eth` / http://offchainexample.eth
3. Verify interstitial to enable Ethereum Name Service is shown.
    - Verify privacy policy and terms of use links open to expected pages for Consensys (Infura)
4. Tap 'Disable'
5. Navigate to Web3 / Wallet settings and verify `Resolve Ethereum Name Service (ENS) Domain Names` is set to `Disabled`
6. Change `Resolve Ethereum Name Service (ENS) Domain Names` back to `Ask`
7. Navigate to any ENS offchain domain via the omnibox. Ex. `offchainexample.eth` / http://offchainexample.eth
8. Tap `Proceed using Infura server`
9. Verify ENS offchain insterstitial is shown.
    - Verify Learn more link opens to the Brave Wallet ENS Offchain privacy considerations page.
10. Tap `Disabled`
11. Navigate to Web3 / Wallet settings and verify `Allow ENS Offchain Lookup` is set to `Disabled`
12. Change `Allow ENS Offchain Lookup` back to `Ask`
13. Navigate to any ENS offchain domain via the omnibox. Ex. `offchainexample.eth` / http://offchainexample.eth
14. Verify offchain insterstitial is shown.
15. Tap `Proceed with offchain lookup`
16. Verify either IPFS interstitial is shown, IPFS public gateway url is loaded, or IPFS disabled page shown (note: IPFS out of scope for this PR).

SNS: http://onybose.sol
ENS: http://offchainexample.eth

## Screenshots:

https://user-images.githubusercontent.com/5314553/227948100-14519772-bafe-41e7-9c8e-63aa8456e1a0.mp4

![Web3 Settings](https://user-images.githubusercontent.com/5314553/227951704-c5511c44-a9aa-475f-9645-893da6ca931f.png) | ![ENS Interstitial](https://user-images.githubusercontent.com/5314553/227951740-2905a220-f1b9-4df7-8646-706c13b8495c.png) | ![ENS Offchain Interstitial](https://user-images.githubusercontent.com/5314553/227951755-60016c95-332f-4c2c-9915-5dc36fc7d7ac.png)
--|--|--


## Reviewer Checklist:

  - `QA/(Yes|No)`
  - `bug` / `enhancement`
